eclass/sys-user.eclass

```diff
@@ -0,0 +1,171 @@
+: ${SYS_USER_HOME:=/dev/null}
+
+# @ECLASS-VARIABLE: SYS_USER_GROUPS
+# @DEFAULT_UNSET
+# @DESCRIPTION:
+# List of groups the user should belong to.  This must be a bash
+# array.  If not specified, the user is not added to any groups.
+# This affects only new user accounts.
+
+
+# << Boilerplate ebuild variables >>
+: ${DESCRIPTION:="System user: ${PN}"}
+: ${HOMEPAGE:=https://www.gentoo.org/}
+: ${LICENSE:=metapackage}
+: ${S:=${WORKDIR}}
+
+
+# << API functions >>
+
+# @FUNCTION: sys-user_add_deps
+# @DESCRIPTION:
+# Generate appropriate DEPEND & RDEPEND from SYS_USER_GROUPS.  This
+# must be called if SYS_USER_GROUPS are set.
+sys-user_add_deps() {
+	debug-print-function ${FUNCNAME} "${@}"
+
+	# SYS_USER_GROUPS sanity check
+	if ! declare -p SYS_USER_GROUPS &>/dev/null; then
+		return
+	elif [[ $(declare -p SYS_USER_GROUPS) != "declare -a"* ]]; then
+		die 'SYS_USER_GROUPS must be an array.'
+	fi
+
+	local g
+	for g in "${SYS_USER_GROUPS[@]}"; do
+		DEPEND+=" group/${g}"
+		RDEPEND+=" group/${g}"
+	done
+
+	_SYS_USER_ADD_DEPS_CALLED=1
+}
+
+
+# << Phase functions >>
+EXPORT_FUNCTIONS pkg_pretend src_install pkg_preinst pkg_prerm
+
+# @FUNCTION: sys-user_pkg_pretend
+# @DESCRIPTION:
+# Performs sanity checks for correct eclass usage, and early-checks
+# whether requested UID can be enforced.
+sys-user_pkg_pretend() {
+	debug-print-function ${FUNCNAME} "${@}"
+
+	# verify that sys-user_add_deps() has been called
+	# (it verifies SYS_USER_GROUPS itself)
+	if [[ -z ${_SYS_USER_ADD_DEPS_CALLED} ]]; then
+		if declare -p SYS_USER_GROUPS &>/dev/null; then
+			die "Ebuild error: sys-user_add_deps must have been called in global scope!"
+		fi
+	fi
+
+	# verify SYS_USER_ID
+	[[ -n ${SYS_USER_ID} ]] || die "Ebuild error: SYS_USER_ID must be set!"
+	[[ ${SYS_USER_ID} -ge 0 ]] || die "Ebuild errors: SYS_USER_ID=${SYS_USER_ID} invalid!"
+
+	# check for SYS_USER_ID collisions early
+	if [[ -n ${SYS_USER_ENFORCE_ID} ]]; then
+		die TODO
+	fi
+}
+
+# @FUNCTION: sys-user_src_install
+# @DESCRIPTION:
+# Installs a keep-file into the user's home directory to ensure it is
+# owned by the package.
+sys-user_src_install() {
+	debug-print-function ${FUNCNAME} "${@}"
+
+	if [[ ${SYS_USER_HOME} != /dev/null ]]; then
+		# note: we can't set permissions here since the user isn't
+		# created yet
+		keepdir "${SYS_USER_HOME}"
+	fi
+}
+
+# @FUNCTION: sys-user_pkg_preinst
+# @DESCRIPTION:
+# Creates the user if it does not exist yet.  Sets permissions
+# of the home directory in install image.
+sys-user_pkg_preinst() {
+	debug-print-function ${FUNCNAME} "${@}"
+
+	local groups=${SYS_USER_GROUPS[*]}
+	enewuser "${PN}" "${SYS_USER_ID}" "${SYS_USER_SHELL}" \
+		"${SYS_USER_HOME}" "${groups// /,}"
+
+	if [[ ${SYS_USER_HOME} != /dev/null ]]; then
+		# set ownership of homedir to user:primary-group
+		fowners "${SYS_USER_ID}" "${SYS_USER_HOME}"
+		if [[ -n ${SYS_USER_GROUPS[0]} ]]; then
+			fowners ":${SYS_USER_GROUPS[0]}" "${SYS_USER_HOME}"
```

Comment:
I think this closes [bug 668180](https://bugs.gentoo.org/show_bug.cgi?id=668180).

Reply:
Or rather, addresses it. The `user.eclass` still won't work right, but we'll be deprecating that I imagine.

Reply:
Exactly. I didn't want to close it as long as the 'faulty' code lives in user.eclass.